### DISCUSSION TITLE
Using Streaming API from non-VF page through jsforce-ajax-proxy

### DISCUSSION THREAD

**Discussion by chexxor** [jsforce/jsforce#1427] [Ideas]
tl;dr: Does JsForce's Streaming API work when using the "jsforce-ajax-proxy" proxy? JsForce is delegating the task to Faye, which doesn't use JsForce's Transport class, so I think this workflow isn't set up yet. Also, I don't know if the "jsforce-ajax-proxy" was designed to proxy messages from the streaming API.

Details:

I'm trying to use JsForce's [Streaming API](http://jsforce.github.io/jsforce/doc/Streaming.html). I'm loading my app from localhost, using OAuth to get an access token, and then sending queries through JsForce. I had to set up a proxy using the [jsforce-ajax-proxy](https://github.com/jsforce/jsforce-ajax-proxy), which works really great.

Now, I'm trying to use the streaming API, but I'm not sure it's possible when using the jsforce-ajax-proxy. I say that because I'm getting this error when Faye is trying to do that handshake. `405 Only POST allowed`

Here's more info about the request which is being sent:

```
Request URL:https://na17.salesforce.com/cometd/30.0?message=%5B%7B%22channel%22%3A%22%2Fmeta%2Fhandshake%22%2C%22version%22%3A%221.0%22%2C%22supportedConnectionTypes%22%3A%5B%22callback-polling%22%5D%2C%22id%22%3A%221%22%7D%5D&jsonp=__jsonp1__
Request Method:GET
Status Code:405 Only POST allowed
Query String Parameters:message:[{"channel":"/meta/handshake","version":"1.0","supportedConnectionTypes":["callback-polling"],"id":"1"}]
jsonp:__jsonp1__
```

Because JsForce is delegating this task of doing the streaming to Faye, the requests sent by Faye won't use JsForce's Transport class, which is where the proxy given to JsForce is used. On top of that, I believe Bayeux protocol is a little complicated and requires server support, so maybe the "jsforce-ajax-proxy" isn't set up to handle JsForce's Streaming API.

Does anyone have info here?

**Comment by chexxor** [comment]
It looks like Faye uses JSON-P to perform the handshake process. JSON-P doesn't uses HTTP headers, while the "jsforce-ajax-proxy" was designed to receive the target URL, a Salesforce resource, via a header, named "salesforceproxy-endpoint".

Perhaps we can patch "jsforce-ajax-proxy" to look for a target URL on the query string as a fallback? I'll give it a try.

Update: I'm still having trouble with this. I'll wait to hear opinions of other people.

**Comment by stomita** [comment]
@chexxor As you commented above, current JSforce doesn't support streaming API from the page resides in external domain. JSforce-ajax-proxy is only for proxying usual http requests and not supporting Bayeux.
According to my former research about streaming API and faye implementation in browser-side, the Salesforce's  Streaming API is not capable for communicating in cross domain. In its protocol handshake they didn't  allow json-p transport. There might be a way to avoid the restriction by proxying or something, but I gave up at that time.

**Comment by chexxor** [comment]
Argh, >_<

"In its protocol handshake they didn't allow json-p transport." - This means Salesforce doesn't respond to json-p?

Thanks for the tips. I think this issue should stay open until someone figures out how to proxy the Streaming API. I might try a few more ideas.

**Comment by Alexander-0x80** [comment]
Hey 👋
I'm facing same issue in 2020
Any chance somebody found a workaround ?
